R/ggridges.R

```diff
@@ -0,0 +1,270 @@
+  )
+}
+
+
+
+#' @export
+to_basic.GeomDensityLine <- function(data, prestats_data, layout, params, p, ...) {
+  to_basic(prefix_class(data, 'GeomDensity'))
+}
+
+
+
+#' @export
+to_basic.GeomRidgeline <- function(data, prestats_data, layout, params, p, ...) {
+  to_basic(
+    prefix_class(data, 'GeomDensityRidgesGradient'),
+    prestats_data, layout, params, p, ...
+  )
+}
+
+
+#' @export
+to_basic.GeomRidgelineGradient <- function(data, prestats_data, layout, params, p, ...) {
+  d <- get_ridge_data(data, params$na.rm)
+  
+  # split data into separate groups
+  groups <- split(d, factor(d$group))
+  
+  # sort list so lowest ymin values are in the front (opposite of ggridges)
+  o <- order(
+    unlist(
+      lapply(
+        groups,
+        function(data) data$ymin[1]
+      )
+    ),
+    decreasing = FALSE
+  )
+  groups <- groups[o]
+  
+  # for each group create a density + vline + point as applicable
+  res <- lapply(
+    rev(groups),
+    function(x){
+      
+      draw_stuff <- split(x, x$datatype)
+      
+      # first draw the basic density ridge part
+      
+      stopifnot(!is.null(draw_stuff$ridgeline))
+      d2 <- d1 <- draw_stuff$ridgeline
+      d2$colour <- NA # no colour for density area
+      d2$fill_plotlyDomain <- NA
+      
+      d1$y <- d1$ymax
+      d1$alpha <- 1 # don't use fill alpha for line alpha
+      
+      # calculate all the positions where the fill type changes
+      fillchange <- c(FALSE, d2$fill[2:nrow(d2)] != d2$fill[1:nrow(d2)-1])
+      
+      # and where the id changes
+      idchange <- c(TRUE, d2$group[2:nrow(d2)] != d2$group[1:nrow(d2)-1])
+      
+      # make new ids from all changes in fill style or original id
+      d2$ids <- cumsum(fillchange | idchange)
+      
+      # get fill color for all ids
+      fill <- d2$fill[fillchange | idchange]
+      
+      # rows to be duplicated
+      dupl_rows <- which(fillchange & !idchange)
+      d2$y <- d2$ymax
+      if (length(dupl_rows)>0){
+        rows <- d2[dupl_rows, ]
+        rows$ids <- d2$ids[dupl_rows-1]
+        rows <- rows[rev(seq_len(nrow(rows))), , drop = FALSE]
+        # combine original and duplicated d2
+        d2 <- rbind(d2, rows)
+      }
+      
+      # split by group to make polygons
+      d2 <- tibble::deframe(tidyr::nest(d2, .by = 'ids'))
+      
+      ridges <- c(
+        d2,
+        list(
+          to_basic(prefix_class(d1, "GeomLine"))
+        )
+      )
+      
+      ridges
+    }
+  )
+  # set list depth to 1
+  unlist(res, recursive = FALSE)
+}
+
+
+
+#' @export
+geom2trace.GeomRidgelineGradient <- function(data, params, p) {
+
+  # munching for polygon
+  positions <- with(data, data.frame(
+    x   = c(x   , rev(x)),
+    y   = c(ymax, rev(ymin))
+  ))
+
+  L <- list(
+    x          = positions[["x"]],
+    y          = positions[["y"]],
+    text       = uniq(data[["hovertext"]]),
+    key        = data[["key"]],
+    customdata = data[["customdata"]],
+    frame      = data[["frame"]],
```

Comment:
I think these need be doubled, like you did for `x`/`y` (this might be leading to https://github.com/plotly/plotly.R/pull/2314#issuecomment-1806071200)

I'd consider doing something like (using `with()` here can be kinda dangerous since it'll error if any of the columns are missing

```r
positions <- data.frame(
  x = c(data$x, rev(data$x)),
  y = c(data$ymax, rev(data$ymin)),
  hovertext = c(data$hovertext, rev(data$hovertext)),
  ...
)
```

Reply:
Found the issue! Within the new `to_basic`s, I was using `to_basic.GeomDensity` and `to_basic.GeomLine` which removed the `crosstalk` set. Attaching them back again  similar to how it's done in `layers2traces` solves the highlight not working. 

Before this, the `to_basic.GeomLine` would preserve the `set` attribute, but `to_basic.GeomDensity` would not, so we were getting highlight working only the line part and not density part. 

I have tested the highlight functionality with different types of ridge plots, with different fills, points, lines within them, and they are all working as expected.

Reply:
@cpsievert are there any pending action items on this, or is this good to go now?

Reply:
Sorry for letting this slip @AdroMine, looking good, would you mind just adding a `NEWS.md` item?

Reply:
@cpsievert , added a line about adding support for ggridges package.